src/runner.rs

```diff
@@ -117,322 +139,274 @@ impl<'a, E: Externals> Interpreter<'a, E> {
-		Ok(InstructionOutcome::RunNextInstruction)
-	}
-
-	fn run_loop(&mut self, context: &mut FunctionContext, labels: &HashMap<usize, usize>, block_type: BlockType) -> Result<InstructionOutcome, TrapKind> {
-		context.push_frame(labels, BlockFrameType::Loop, block_type)?;
-		Ok(InstructionOutcome::RunNextInstruction)
+	fn run_br(&mut self, _context: &mut FunctionContext, target: isa::Target) -> Result<InstructionOutcome, TrapKind> {
+		Ok(InstructionOutcome::Branch(target))
 	}
 
-	fn run_if(&mut self, context: &mut FunctionContext, labels: &HashMap<usize, usize>, block_type: BlockType) -> Result<InstructionOutcome, TrapKind> {
-		let condition: bool = context
-			.value_stack_mut()
-			.pop_as();
-		let block_frame_type = if condition { BlockFrameType::IfTrue } else {
-			let else_pos = labels[&context.position];
-			if !labels.contains_key(&else_pos) {
-				context.position = else_pos;
-				return Ok(InstructionOutcome::RunNextInstruction);
-			}
-
-			context.position = else_pos;
-			BlockFrameType::IfFalse
-		};
-		context.push_frame(labels, block_frame_type, block_type)?;
-
-		Ok(InstructionOutcome::RunNextInstruction)
-	}
-
-	fn run_else(&mut self, context: &mut FunctionContext, labels: &HashMap<usize, usize>) -> Result<InstructionOutcome, TrapKind> {
-		let end_pos = labels[&context.position];
-		context.pop_frame(false)?;
-		context.position = end_pos;
-		Ok(InstructionOutcome::RunNextInstruction)
-	}
-
-	fn run_end(&mut self, context: &mut FunctionContext) -> Result<InstructionOutcome, TrapKind> {
-		context.pop_frame(false)?;
-		Ok(InstructionOutcome::End)
-	}
-
-	fn run_br(&mut self, _context: &mut FunctionContext, label_idx: u32) -> Result<InstructionOutcome, TrapKind> {
-		Ok(InstructionOutcome::Branch(label_idx as usize))
+	fn run_br_nez(&mut self, target: isa::Target) -> Result<InstructionOutcome, TrapKind> {
+		let condition = self.value_stack.pop_as();
+		if condition {
+			Ok(InstructionOutcome::Branch(target))
+		} else {
+			Ok(InstructionOutcome::RunNextInstruction)
+		}
 	}
 
-	fn run_br_if(&mut self, context: &mut FunctionContext, label_idx: u32) -> Result<InstructionOutcome, TrapKind> {
-		let condition = context.value_stack_mut().pop_as();
+	fn run_br_eqz(&mut self, target: isa::Target) -> Result<InstructionOutcome, TrapKind> {
+		let condition = self.value_stack.pop_as();
 		if condition {
-			Ok(InstructionOutcome::Branch(label_idx as usize))
-		} else {
 			Ok(InstructionOutcome::RunNextInstruction)
+		} else {
+
+			Ok(InstructionOutcome::Branch(target))
 		}
 	}
 
-	fn run_br_table(&mut self, context: &mut FunctionContext, table: &[u32], default: u32) -> Result<InstructionOutcome, TrapKind> {
-		let index: u32 = context.value_stack_mut()
+	fn run_br_table(&mut self, table: &[isa::Target]) -> Result<InstructionOutcome, TrapKind> {
+		let index: u32 = self.value_stack
 			.pop_as();
-		Ok(InstructionOutcome::Branch(table.get(index as usize).cloned().unwrap_or(default) as usize))
+
+		let dst =
```

Comment:
strange identation below

Reply:
Maybe we can use `rustfmt` from here on?

Reply:
I'd like to, but let's postpone it until the next PR!